### DISCUSSION TITLE
Packing A5Voice segment to send to Hydrasynth

### DISCUSSION THREAD

**Discussion by mungewell** [patois/A5Pack#9] [General]
I think that 'we' are at the point where we have an altered A5Voice segment ready to send back to the Hydrasynth, and I presume that this needs to be packed into a '.dat' file - how is this done?

I have uploaded my script `A5Voice.py` to a branch on my fork:
https://github.com/mungewell/A5Pack/tree/A5Voice_v1

This includes an example bash script, which:

- exports the samples for waveform 1 (Sine)
- renames them suitably
- imports sample so that they overwrite waveform2 (which was Triangle)

This results in a new '.bin' file:
```
~/A5Pack-sdw-github/A5Voice$ md5sum *.bin
91fbb6eeca196476fe6d3a37a2668665  a5_voice_0x01600000.bin
dc6b58089229685ae2a5ed455b3eae8d  messed_with.bin
```

I also confirmed that the script works OK in Windows, but example will need to be converted to `.bat`.

**Comment by patois** [comment]
Nice, it looks as if you're really close to figuring out how to have custom wavetables on the HS :)

As for your question: I haven't looked into any of it personally but there are two options right now:
* taking a modified A5Voice section and manually pasting it over the section of a legit .dat file with a hex editor, then flashing the modified .dat file (this is basically what I did in the past when I've replaced the text that now says "Pwned!", as can be seen in the screenshot on this project. By having succeeded in doing so, it was proven that there is no checksum/integrity check in place, as in: patching data/code is "relatively safe")
* if I recall correctly, using the development version of the HydraSynth firmware updater, a custom firmware update can be built from individual sections, which *should* result in a valid firmware image (check "edit5" in my original post on reddit: https://www.reddit.com/r/hydrasynth/comments/11uhyjx/reverse_engineering_and_customizing_hydrasynth/)

Please note that I do not recommend flashing any modified firmware without willingly taking the risk of potentially turning your HS into a brick.

A third option would be reverse engineering the updater (if that's a legal thing to do in the country of your residence) and figuring out how to create a valid firmware image from individual sections and implementing a custom tool for rebuilding an unpacked firmware image. This had originally been the goal when development of A5Pack started (hence the name) but repacking functionality hasn't been implemented because it has kind of been obsoleted with the release of the development version of the HS updater by Ashun Sound Machines.

Now since A5Pack has been added an MIT license, please feel free to work on the fork of yours instead - the original A5Pack project will probably be no longer be maintained by myself any longer.

Good luck and happy hacking!

**Reply by mungewell** [reply]
I had missed that the dev-tool allowed the creation of '.dat' packs, so I built for the various devices. Whilst 'we' have that at present, others may not be able to acquire in the future...

I note that the md5sum are not the same.
```
$ cat md5sum.txt 
91fbb6eeca196476fe6d3a37a2668665 *A5_voice.bin
8e22b330bc31e96a85c1a34f6868b156 *A5_voice_delux.dat
9ea82af4804d45fe5dfe49cffd2e67f0 *A5_voice_desktop.dat
c26daa431145a0ad77927a46b841655e *A5_voice_explorer.dat
9c1f3271fa50799e0d18947e673e6bff *A5_voice_keyboard.dat
```

This is because each pack is prefixed specially for each keyboard
```
==> delux.txt <==
00000000  41 35 50 61 63 6b 00 37  58 00 00 41 6c 6c 20 52  |A5Pack.7X..All R|
00000010  69 67 68 74 73 20 52 65  73 65 72 76 65 64 00 00  |ights Reserved..|

==> desktop.txt <==
00000000  41 35 50 61 63 6b 00 44  52 00 00 41 6c 6c 20 52  |A5Pack.DR..All R|
00000010  69 67 68 74 73 20 52 65  73 65 72 76 65 64 00 00  |ights Reserved..|

==> explorer.txt <==
00000000  41 35 50 61 63 6b 00 33  4d 00 00 41 6c 6c 20 52  |A5Pack.3M..All R|
00000010  69 67 68 74 73 20 52 65  73 65 72 76 65 64 00 00  |ights Reserved..|

==> keyboard.txt <==
00000000  41 35 50 61 63 6b 00 4b  42 00 00 41 6c 6c 20 52  |A5Pack.KB..All R|
00000010  69 67 68 74 73 20 52 65  73 65 72 76 65 64 00 00  |ights Reserved..|
```

I'd still like confirmation that my understanding of the 'A5 Voice' segment is correct. I suspect that the Explorer has a reduced memory size, so that might mean that the segment has a different size/formatting. _None of the official FW contain the 'A5 Voice' segment._

**Reply by patois** [reply]
I'm afraid I am unable to confirm your understanding of the "A5 Voice" segment because I don't know its format. I stopped reverse engineering the HS firmware and updaters myself due to lack of time but you should be able to figure it out if you happen to know the C programming language. I'd grab a copy of the free decompiler Ghidra (https://ghidra-sre.org/) and have a look at the development version of the updater(s). I'd assume that finding a "A5 Voice" string and cross referencing it will probably lead to code that processes it - otherwise, if the string is not present, I'd assume the "A5 Voice" segment can't be flashed. Another approach would be building a firmware that contains an "A5 Voice" segment using the development tool by ASM and having a look at the resulting file with a hexeditor, comparing it to other firmware images and unpacking it again using either the dev tool or A5Pack in order to find out whether it looks sane.

Last but not least, I'd like to ask you a favor: if you could please refrain from pasting any of the ASM copyrighted material (firmware/decompiled code/data) in order to stay in a legal frame.

Thank you and good luck.

**Comment by mungewell** [comment]
The point of no return....
![point_of_no_return](https://github.com/patois/A5Pack/assets/236907/92c2fc23-700b-4d1e-b8b6-d80571cae5c0)

And where it ended. :-)
https://youtu.be/PW-fqcufOTQ
